Panic when returning a nil func

The following program `sample.go` triggers a panic:

```go
package main

import (
	"fmt"

	"github.com/containous/yaegi/interp"
)

const code = `
package foo

func Bar() func() {
	return nil
}
`

func main() {
	i := interp.New(interp.Options{})

	_, err := i.Eval(code)
	if err != nil {
		panic(err)
	}

	v, err := i.Eval("foo.Bar")
	if err != nil {
		panic(err)
	}

	f, ok := v.Interface().(func() func())
	if !ok {
		panic("unable to convert value")
	}

	r := Bar()
	fmt.Printf("got function %v\n", r)
	r = f()
	fmt.Printf("got function %v\n", r)
}

func Bar() func() {
	return nil
}

```

Expected result:
```console
$ go run ./sample.go
// output
got function <nil>
got function <nil>
```

Got:
```console
$ go run ./sample.go
// output
got function <nil>
5:2: panic
panic: reflect.Set: value of type func() is not assignable to type *interp.node [recovered]
	panic: reflect.Set: value of type func() is not assignable to type *interp.node

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc0000be3c0, 0xc0000c6e00)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:112 +0x218
panic(0x5ff760, 0xc0000808a0)
	/usr/lib/go/src/runtime/panic.go:679 +0x1b2
reflect.Value.assignTo(0x5fcfc0, 0xc0000a00a8, 0x193, 0x656d30, 0xb, 0x6327a0, 0x0, 0x12ca0, 0x5f2ca0, 0xc0000a0138)
	/usr/lib/go/src/reflect/value.go:2403 +0x432
reflect.Value.Set(0x6327a0, 0xc0000a0138, 0x196, 0x5fcfc0, 0xc0000a00a8, 0x193)
	/usr/lib/go/src/reflect/value.go:1537 +0xbd
github.com/containous/yaegi/interp._return.func3(0xc0000be3c0, 0x6327a0)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:1338 +0x8a
github.com/containous/yaegi/interp.runCfg(0xc0000c6e00, 0xc0000be3c0)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:118 +0x7a
github.com/containous/yaegi/interp.genFunctionWrapper.func1.1(0x80d820, 0x0, 0x0, 0x80d820, 0x657e20, 0x1)
	/home/jlevesy/code/src/github.com/containous/yaegi/interp/run.go:482 +0x3a5
reflect.callReflect(0xc000082e10, 0xc00007ded0, 0xc00007deb8)
	/usr/lib/go/src/reflect/value.go:549 +0x322
reflect.makeFuncStub(0x699760, 0xc0000a0008, 0x657e11, 0x10, 0xc00007df40, 0x1, 0x1, 0x13, 0x0, 0x0, ...)
	/usr/lib/go/src/reflect/asm_amd64.s:20 +0x42
main.main()
	/home/jlevesy/code/src/experiments/yaegi/case1/main.go:37 +0x190
exit status 2
```

--- a/_test/fun10.go
+++ b/_test/fun10.go
@@ -1,0 +1,19 @@
+package main
+
+import "fmt"
+
+func f() func() {
+	return nil
+}
+
+func main() {
+	g := f()
+	fmt.Printf("%T %v\n", g, g)
+	if g == nil {
+		fmt.Println("nil func")
+	}
+}
+
+// Output:
+// func() <nil>
+// nil func

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1035,7 +1035,12 @@ func (interp *Interpreter) cfg(root *node) ([]*node, error) {
 					if typ, err = nodeType(interp, sc, f.child[2].child[1].child[i].lastChild()); err != nil {
 						return
 					}
-					c.rval = reflect.New(typ.TypeOf()).Elem()
+					if typ.cat == funcT {
+						// Wrap the typed nil value in a node, as per other interpreter functions
+						c.rval = reflect.ValueOf(&node{kind: basicLit, rval: reflect.New(typ.TypeOf()).Elem()})
+					} else {
+						c.rval = reflect.New(typ.TypeOf()).Elem()
+					}
 				}
 			}
 

--- a/interp/interp_eval_test.go
+++ b/interp/interp_eval_test.go
@@ -156,6 +156,14 @@ func TestEvalNil(t *testing.T) {
 			src: "Hello()",
 			res: "<nil>",
 		},
+		{
+			desc: "return nil func",
+			pre: func() {
+				eval(t, i, `func Bar() func() { return nil }`)
+			},
+			src: "Bar()",
+			res: "<nil>",
+		},
 	})
 }
 

--- a/interp/run.go
+++ b/interp/run.go
@@ -428,6 +428,10 @@ func _panic(n *node) {
 func genFunctionWrapper(n *node) func(*frame) reflect.Value {
 	var def *node
 	var ok bool
+
+	if n.kind == basicLit {
+		return func(f *frame) reflect.Value { return n.rval }
+	}
 	if def, ok = n.val.(*node); !ok {
 		return genValueAsFunctionWrapper(n)
 	}
@@ -2359,7 +2363,12 @@ func slice0(n *node) {
 }
 
 func isNil(n *node) {
-	value := genValue(n.child[0])
+	var value func(*frame) reflect.Value
+	if n.child[0].typ.cat == funcT {
+		value = genValueAsFunctionWrapper(n.child[0])
+	} else {
+		value = genValue(n.child[0])
+	}
 	tnext := getExec(n.tnext)
 
 	if n.fnext != nil {
@@ -2380,7 +2389,12 @@ func isNil(n *node) {
 }
 
 func isNotNil(n *node) {
-	value := genValue(n.child[0])
+	var value func(*frame) reflect.Value
+	if n.child[0].typ.cat == funcT {
+		value = genValueAsFunctionWrapper(n.child[0])
+	} else {
+		value = genValue(n.child[0])
+	}
 	tnext := getExec(n.tnext)
 
 	if n.fnext != nil {

--- a/interp/run.go
+++ b/interp/run.go
@@ -1354,6 +1354,8 @@ func _return(n *node) {
 			} else {
 				values[i] = genValue(c)
 			}
+		case funcT:
+			values[i] = genValue(c)
 		case interfaceT:
 			values[i] = genValueInterface(c)
 		default:

--- a/interp/value.go
+++ b/interp/value.go
@@ -55,9 +55,15 @@ func genValueRecv(n *node) func(*frame) reflect.Value {
 }
 
 func genValueAsFunctionWrapper(n *node) func(*frame) reflect.Value {
-	v := genValue(n)
+	value := genValue(n)
+	typ := n.typ.TypeOf()
+
 	return func(f *frame) reflect.Value {
-		return genFunctionWrapper(v(f).Interface().(*node))(f)
+		v := value(f)
+		if v.IsNil() {
+			return reflect.New(typ).Elem()
+		}
+		return genFunctionWrapper(v.Interface().(*node))(f)
 	}
 }
 

